src/libcore/iter/adapters/zip.rs

```diff
@@ -295,19 +324,105 @@ where
 {
 }
 
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A: Debug, B: Debug> Debug for Zip<A, B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        ZipFmt::fmt(self, f)
+    }
+}
+
+trait ZipFmt<A, B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result;
+}
+
+impl<A: Debug, B: Debug> ZipFmt<A, B> for Zip<A, B> {
+    default fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Zip").field("a", &self.a).field("b", &self.b).finish()
+    }
+}
+
+impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B> for Zip<A, B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // It's *not safe* to call fmt on the contained iterators, since once
+        // we start iterating they're in a strange potentially unsafe states.
+        f.debug_struct("Zip").finish()
+    }
+}
+
 /// An iterator whose items are random-accessible efficiently
 ///
 /// # Safety
 ///
-/// The iterator's .len() and size_hint() must be exact.
-/// `.len()` must be cheap to call.
+/// The iterator's `size_hint` must be exact and cheap to call.
+///
+/// `size` may not be overridden.
+///
+/// `<Self as Iterator>::get_unchecked` must be overridden. It must be safe to
+/// call provided the following conditions are met
 ///
-/// .get_unchecked() must return distinct mutable references for distinct
-/// indices (if applicable), and must return a valid reference if index is in
-/// 0..self.len().
-pub(crate) unsafe trait TrustedRandomAccess: ExactSizeIterator {
-    unsafe fn get_unchecked(&mut self, i: usize) -> Self::Item;
+/// 1. `0 <= idx` and `idx < self.size()`.
+/// 2. If `self: !Clone`, then `get_unchecked` is never called with the same
+///    index on `self` more than once.
+/// 3. After `self.get_unchecked(idx)` has been called then `next_back` will
+///    only be called at most `self.size() - idx - 1` times.
+/// 4. After `get_unchecked` is called, then only the following methods will be
+///    called on `self`:
+///     * `std::clone::Clone::clone`
+///     * `std::iter::Iterator::size_hint()`
+///     * `std::iter::Iterator::next_back()`
+///     * `std::iter::Iterator::get_unchecked()`
+///     * `std::iter::TrustedRandomAccess::size()`
+///
+/// Further, given that these conditions are met, it must guarantee that:
+///
+/// * It does not change the value returned from `size_hint`
+/// * It must be safe to call the methods listed above on `self` after calling
+///   `get_unchecked`, assuming that the required traits are implemented.
+/// * It must also be safe to drop `self` after calling `get_unchecked`.
+#[doc(hidden)]
+#[unstable(feature = "trusted_random_access", issue = "none")]
+#[rustc_specialization_trait]
+pub unsafe trait TrustedRandomAccess: Sized {
```

Comment:
Why does this need to become `pub`?

Reply:
It's now used in the interface of `Iterator`